FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py

```diff
@@ -0,0 +1,289 @@
+        return (value.nominal_value, value.std_dev)
+
+
+@dataclass 
+class MadXBoundaryConditions:
+    """Store all boundary conditions for a Mad-X twiss."""
+    alfx: float = None
+    alfy: float = None
+    betx: float = None
+    bety: float = None
+    dx: float = None
+    dy: float = None
+    dpx: float = None
+    dpy: float = None
+    wx: float = None
+    wy: float = None
+    dphix: float = None
+    dphiy: float = None
+    r11: float = None
+    r12: float = None
+    r21: float = None
+    r22: float = None
+
+    def as_dict(self):
+        return {f.name: getattr(self, f.name) for f in fields(self) if getattr(self, f.name) is not None}
+
+
+# Main -------------------------------------------------------------------------
+
+def create_realizations(output: Path, n: int = 100) -> pd.DataFrame:
+    """Main function to generate the phase advance realizations.
+    The mean and std values are hardcoded and choosen to have a final phase advance 
+    of around 0.25 (far away from 0.5) in units of 2 Pi.
+
+    Args:
+        output (Path): Path to save the output tfs-file.
+        n (int, optional): Number of realizations to generate. Defaults to 100.
+    """
+    madx = generate_fodo()
+    madx.twiss(chrom=True)
+
+    x = PropagableBoundaryConditions(alpha=ufloat(-1.6, 0.002), beta=ufloat(3.467, 0.025))
+    y = PropagableBoundaryConditions(alpha=ufloat(2.16, 0.002), beta=ufloat(7.14, 0.016))
+    no_error = MadXBoundaryConditions(
+        alfx=x.alpha.nominal_value, 
+        betx=x.beta.nominal_value, 
+        alfy=y.alpha.nominal_value, 
+        bety=y.beta.nominal_value,
+    )
+    
+    df_ref = generate_df(madx, sequence="fodo", boundary_conditions=no_error)
+    
+    df = pd.DataFrame()
+    df["MUX"] = df_ref.mux
+    df["PLUSX"] = propagate_with_plus(dphi=df_ref.mux, init=x)
+    df["MINUSX"] = propagate_with_minus(dphi=df_ref.mux, init=x)
+    
+    df["MUY"] = df_ref.muy
+    df["PLUSY"] = propagate_with_plus(dphi=df_ref.muy, init=y)
+    df["MINUSY"] = propagate_with_minus(dphi=df_ref.muy, init=y)
+    print(df_ref[["mux", "muy"]])
+
+    for idx in range(n):
+        with_error = MadXBoundaryConditions(
+            alfx=get_realization(x.alpha),
+            betx=get_realization(x.beta),
+            alfy=get_realization(y.alpha),
+            bety=get_realization(y.beta),
+        )
+
+        df_err = generate_df(madx, sequence="fodo", boundary_conditions=with_error)
+        
+        diff_x = (df_err.mux - df_ref.mux) % 1
+        diff_y = (df_err.muy - df_ref.muy) % 1
+        diff_x[diff_x > 0.5] -= 1
+        diff_y[diff_y > 0.5] -= 1
+
+        df[f"MADXX{idx:03d}"] = diff_x
+        df[f"MADXY{idx:03d}"] = diff_y
+
+    tfs.write(output, df, save_index="NAME")
+    return df
+
+
+# Helper ---
+
+def generate_fodo(sequence: str ="fodo") -> Madx:
+    """Generate a standard FODO sequence.
+    
+    Args:
+        sequence (str): Sequence name.
+    """
+    madx = Madx()
```

Comment:
for general (re)use, wouldn't it be better to provide the cpyMAD-X instance as input to the fct?

Reply:
and for more fun studies, the phase adv. of the cell could also be an input, to compare e.g. 60 and 90 deg LHC stuff